feat(transformer): complete the async-to-generator plugin

Dunqing opened the pull request:
In this PR, most of the async functions have transformed correctly. But the async arrow functions don't fully transform correctly yet, it is related to we need to transform the arrow function to the generator function. For example:

Input:
```js
function declaration() {
  const asy = async () => {
  console.log(this.name)
 }
}
```

Output:
```js
function declaration() {
  const asy = babelHelpers.asyncToGenerator(function* () {
     console.log(this.name);
  });
}
```

Expected Output:

```js
function declaration() {
  var _this = this;
  const asy = /*#__PURE__*/function () {
    var _ref = babelHelpers.asyncToGenerator(function* () {
      console.log(_this.name);
    });
    return function asy() {
      return _ref.apply(this, arguments);
    };
  }();
}
```

From the expected output, we haven't handled `this` correctly, which means even if the `arrow-function` plugin doesn't enable, we still need to handle this correctly as the `arrow-function` plugin does, and further question if `arrow-function` plugin is enabled, how to avoid these making conflict? 

I thought we may move out the implementation of `arrow-function` and as a common helper, this way every plugin can handle this well

Dunqing commented:
* **#6658** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6658?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @Dunqing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented on crates/oxc_semantic/src/scope.rs:
```diff
@@ -173,6 +173,16 @@ impl ScopeTree {
     }
 
+    pub fn change_parent_id(&mut self, scope_id: ScopeId, new_parent_id: Option<ScopeId>) {
+        let old_parent_id = self.parent_ids[scope_id];
+        if self.build_child_ids {
+            if let Some(old_parent_id) = old_parent_id {
+                self.child_ids[old_parent_id].retain(|&child_id| child_id != scope_id);
+            }
+        }
```
I think need to record this scope as child of its new parent too.

```suggestion
        if self.build_child_ids {
            let old_parent_id = self.parent_ids[scope_id];
            if let Some(old_parent_id) = old_parent_id {
                self.child_ids[old_parent_id].retain(|&child_id| child_id != scope_id);
            }
            if let Some(new_parent_id) = new_parent_id {
                self.child_ids[new_parent_id].push(scope_id);
            }
        }
```

Dunqing replied:
Yes, it was handled in `self.set_parent_id`. But it causes two repeated branch checks for `self.build_child_ids`. I will refactor it so that it doesn't call `self.set_parent_id` anymore

Dunqing commented on crates/oxc_transformer/src/common/statement_injector.rs:
```diff
@@ -131,6 +130,8 @@ impl<'a> StatementInjectorStore<'a> {
                     new_statements.push(stmt);
                     new_statements.extend(adjacent_stmts.into_iter().map(|s| s.stmt));
                 }
+            } else {
+                new_statements.push(stmt);
```
Good catch! I didn't realize this

overlookmotel commented:
Oh I'm sorry. #6856 or #6858 has caused merge conflicts and likely broken things. I should have reviewed and merged this PR first.

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
```diff
@@ -17,16 +17,28 @@
-//!   return _asyncToGenerator(function* () {
-//!     yield bar();
-//!   })
+//!        return _foo.apply(this, arguments);
```
Nit: The indentation in these examples seems a bit wacky/excessive.

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
```diff
@@ -1,6 +1,6 @@
 //! ES2017: Async / Await \[WIP\]
 //!
-//! This plugin transforms async functions to generator functions.
+//! This plugin transforms async functions to generator functions and wraps them in the asyncToGenerator helper function.
```
Style nit!

```suggestion
//! This plugin transforms async functions to generator functions
//! and wraps them with `asyncToGenerator` helper function.
```

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
```diff
@@ -35,21 +47,18 @@
+use std::mem;
+
+use oxc_allocator::{Box, GetAddress};
+use oxc_ast::ast::*;
+use oxc_ast::Visit;
+use oxc_ast::NONE;
```
Another style nit. Don't we usually go with this style?

```suggestion
use oxc_ast::{ast::*, Visit, NONE};
```

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
```diff
@@ -65,138 +74,318 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
+    fn transform_await_expression(
         &self,
-        func: &mut Function<'a>,
+        expr: &mut Box<'a, AwaitExpression<'a>>,
```
`&mut Box<'a, AwaitExpression<'a>>` is a double-reference (`&mut` is a reference, and so is `Box`). It might make more sense to take a `&mut AwaitExpression<'a>`.

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
````diff
@@ -65,138 +74,318 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
+    /// ```
+    fn transform_function_for_method_definition(
+        &self,
+        func: &mut Box<'a, Function<'a>>,
````
Same as above. Probably better to take a `&mut Function<'a>`.

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
```diff
@@ -65,138 +74,318 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
+    /// and wraps it in asyncToGenerator helper function.
+    fn transform_function_expression(
+        &self,
+        wrapper_function: &mut Box<'a, Function<'a>>,
```
Same as above. Probably better to take a `&mut Function<'a>`.

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
```diff
@@ -65,138 +74,318 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
+    /// Transforms async function declarations into generator functions wrapped in the asyncToGenerator helper.
+    fn transform_function_declaration(
+        &self,
+        wrapper_function: &mut Box<'a, Function<'a>>,
```
Same as above. Probably better to take a `&mut Function<'a>`.

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
```diff
@@ -208,24 +397,306 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
+    }
+
+    /// Infers the function id from [`Ancestor::VariableDeclaratorInit`].
+    #[inline]
```
What's the reason for `#[inline]` here?

Generally it's better to leave decisions on whether to inline or not to the compiler, except when:

1. the function is trivial (only a couple of assembly instructions) or
2. there's some specific circumstances where you know more than the compiler, or want to "tell" the compiler something.

An example of (2) is when the function needs to be inlined for the compiler to prove an `unwrap()` will never panic and that it can remove it:

```rs
fn caller(expr: &Expression) {
    if matches!(expr, Expression::BinaryExpression(_) {
        // `expr` is definitely a `BinaryExpression`
        callee(expr);
    }
}

// `#[inline]` so that compiler can prove that `expr` is always a `BinaryExpression`
// and remove the `unreachable!()` branch
#[inline]
fn callee(expr: &Expression) {
    // If this function isn't inlined, the compiler knows nothing about the type of `expr`
    // because each function is optimized separately.
    // But if this function is inlined into `caller`, then compiler sees all the code together.
    // It knows `expr` is a `BinaryExpression` because of the `if` statement in `caller`,
    // so it will remove the `else { unreachable!() }` branch here.
    let Expression::BinaryExpression(bin_expr) = expr else { unreachable!() };
}
```

If you're doing something like this, include a comment to explain why `#[inline]` (like the comment above `callee` in example above).

But my rule of thumb is that if I'm not sure whether to inline or not, don't. I trust the compiler more than myself!

Same for most of the other functions below.

Dunqing replied:
Thanks for explaining a lot, I understand now!

overlookmotel commented on crates/oxc_transformer/src/es2017/async_to_generator.rs:
```diff
@@ -208,24 +397,306 @@ impl<'a, 'ctx> AsyncToGenerator<'a, 'ctx> {
+    }
+
+    /// Creates a [`Function`] with the specified params, body and scope_id.
+    #[inline]
```
`#[inline]` seems sensible here. This function just constructs a type, and inlining it makes it more likely that compiler will see it can construct the type directly in arena, rather than constructing on stack and then moving to the arena.

overlookmotel reviewed (approved):
I've not reviewed the logic of this transform, as I don't know it well enough. But this PR makes the tests pass so 👍!

All my comments are style nits, and some notes about use of `#[inline]`. I hope that's useful, and not annoying!

Dunqing commented:
Mering this!

Dunqing commented:
### Merge activity

* **Oct 24, 11:18 PM EDT**: The merge label '0-merge' was detected. This PR will be added to the [Graphite merge queue](https://app.graphite.dev/merges?org=oxc-project&repo=oxc) once it meets the requirements.

Dunqing committed:
feat(transformer): complete the async-to-generator plugin (#6658)

In this PR, most of the async functions have transformed correctly. But the async arrow functions don't fully transform correctly yet, it is related to we need to transform the arrow function to the generator function. For example:

Input:
```js
function declaration() {
  const asy = async () => {
  console.log(this.name)
 }
}
```

Output:
```js
function declaration() {
  const asy = babelHelpers.asyncToGenerator(function* () {
     console.log(this.name);
  });
}
```

Expected Output:

```js
function declaration() {
  var _this = this;
  const asy = /*#__PURE__*/function () {
    var _ref = babelHelpers.asyncToGenerator(function* () {
      console.log(_this.name);
    });
    return function asy() {
      return _ref.apply(this, arguments);
    };
  }();
}
```

From the expected output, we haven't handled `this` correctly, which means even if the `arrow-function` plugin doesn't enable, we still need to handle this correctly as the `arrow-function` plugin does, and further question if `arrow-function` plugin is enabled, how to avoid these making conflict?

I thought we may move out the implementation of `arrow-function` and as a common helper, this way every plugin can handle this well
